RandLAPACK/comps/rl_util.hh

```diff
@@ -370,6 +370,189 @@ void gen_exp_mat(
     }
 }
 
+/// Generates matrix with a staircase spectrum with 4 steps
+template <typename T>
+void gen_step_mat(
+    int64_t& m,
+    int64_t& n,
+    std::vector<T>& A,
+    int64_t k,
+    T cond,
+    bool diagon,
+    RandBLAS::base::RNGState<r123::Philox4x32> state
+) {
+
+    // Predeclare to all nonzero constants, start decay where needed
+    std::vector<T> s(k, 1.0);
+    std::vector<T> S(k * k, 0.0);
+
+    // We will have 4 steps controlled by the condition number size and starting with 1
+    int offset = (int) (k / 4);
+
+    std::fill(s.begin(), s.begin() + offset, 1);
+    std::fill(s.begin() + offset + 1, s.begin() + 2 * offset, 8.0 / cond);
+    std::fill(s.begin() + 2 * offset + 1, s.begin() + 3 * offset, 4.0 / cond);
+    std::fill(s.begin() + 3 * offset + 1, s.end(), 1.0 / cond);
+
+    // form a diagonal S
+    diag(k, k, s, k, S);
+
+    if (diagon) {
+        if (!(m == k || n == k)) {
+            m = k;
+            n = k;
+            A.resize(k * k);
+        }
+        lapack::lacpy(MatrixType::General, k, k, S.data(), k, A.data(), k);
+    } else {
+        gen_mat(m, n, A, k, S, state);
+    }
+}
+
+/// Generates a matrix with high coherence between the left singular vectors.
+/// Such matrix would be difficult to sketch.
+template <typename T>
+void gen_spiked_mat(
+    int64_t& m,
+    int64_t& n,
+    std::vector<T>& A,
+    RandBLAS::base::RNGState<r123::Philox4x32> state
+) {
+
+    T* A_dat = A.data();
+
+    int64_t num_rows_sampled = n / 2;
+
+    RandBLAS::sparse::SparseDist DS = {RandBLAS::sparse::SparseDistName::LASO, num_rows_sampled, m, 1};
+    RandBLAS::sparse::SparseSkOp<T> S(DS, state, NULL, NULL, NULL);
+    RandBLAS::sparse::fill_sparse(S);
+
+    // Fill A with stacked identities
+    int start = 0;
+    while(start + n <= m){
+        for(int j = 0; j < n; ++j) {
+            A_dat[start + (m * j) + j] = 1.0;
+        }
+        start += n;
+    }
+
+    // Scale randomly sampled rows
+    start = 0;
+    while (start + m <= m * n) {
+        for(int i = 0; i < num_rows_sampled; ++i) {
+            A_dat[start + (S.cols)[i] - 1] *= m;
```

Comment:
Please have this function accept a proportionality parameter ``prop_n`` so that ``num_rows_sampled = prop_n * n`` (right now, you've hard-coded ``prop_n = 0.5``).

Please also have it accept a parameter ``spike_scale`` that controls the scale of the spiked rows (right now, you have a hard-coded factor of ``m``).

Reply:
Fixed the latter. We should discuss the former.

Reply:
Per your comment on Slack

> The problem is that the API for the matrix generation now only has 3 parameters that the user can specify: mat type, an arbitrary parameter and whether they want a diagonal matrix. We're already using an arbitrary one for the scaling factor. There are ways to make your idea work, but that requires an API change, which means fixing it all around RandLAPACK.

I'm fine leaving the latter unaddressed.